proof/refine/X64/Retype_R.thy

```diff
@@ -4374,17 +4374,14 @@ lemma createNewCaps_idle'[wp]:
                       split del: if_split)
          apply (rename_tac apiobject_type)
          apply (case_tac apiobject_type, simp_all split del: if_split)[1]
-             apply (wp, simp)
-           including classic_wp_pre
-           apply (wp mapM_x_wp'
-                     createObjects_idle'
-                     threadSet_idle'
+             apply wpsimp
+            apply (wpsimp wp: mapM_x_wp' createObjects_idle' threadSet_idle'
                    | simp add: projectKO_opt_tcb projectKO_opt_cte
                                makeObject_cte makeObject_tcb archObjSize_def
                                tcb_cte_cases_def objBitsKO_def APIType_capBits_def
                                objBits_def createObjects_def bit_simps
                    | intro conjI impI
-                   | fastforce simp: curDomain_def)+
+                   | clarsimp simp: curDomain_def)+
```

Comment:
Is this a consequence of removing `hoare_gets`, or was fastforce more powerful than was needed in general? I've noticed a few instances of that in this commit.

Reply:
It's a consequence of removing `hoare_gets`. Roughly, this proof previously used `wp` until it hit `curDomain`, which the `fastforce` would unfold and then solve using `hoare_gets` as an `intro` rule. Now, the `clarsimp` is just being used to do the unfold and then it goes back to `wp` to solve the goal. I don't remember exactly why, but for some reason I thought that the unfold couldn't happen ahead of time.

Reply:
Looking at it again here, the `clarsimp` and `simp` could probably be combined though.